File: java/structure/linear/linked_list/CircularLinkedList.java
package structure.linear.linked_list;

public class CircularLinkedList {
    public boolean isCircular(ListNode head) {
        ListNode slow = head.next;
        ListNode fast = head.next;
        while (null != fast && null != fast.next) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }
        return false;
    }

    public int circularLength(ListNode head) {
        ListNode slow = head.next;
        ListNode fast = head.next;
        while (null != fast && null != fast.next) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                break;
            }
        }

        if (null == fast || null == fast.next) {
            return 0;
        }

        slow = slow.next;
        int count = 1;
        while (slow != fast) {
            slow = slow.next;
            ++count;
        }
        return count;
    }

    public ListNode circularEntry(ListNode head) {
        ListNode slow = head.next;
        ListNode fast = head.next;
        while (null != fast && null != fast.next) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                break;
            }
        }

        if (null == fast || null == fast.next) {
            return null;
        }

        fast = head.next;
        while (slow != fast) {
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }

    public class ListNode {
        public int value = 0;
        public ListNode next = null;

        public ListNode(int value) {
            this.value = value;
        }
    }
}

File: py/structure/linear/linked_list/CircularLinkedList.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class ListNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class CircularLinkedList:
    def isCircular(self, head: ListNode) -> bool:
        slow, fast = head.next, head.next
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow == fast:
                return True
        return False

    def circularLength(self, head: ListNode) -> int:
        slow, fast = head.next, head.next
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow == fast:
                break

        if not fast or not fast.next:
            return 0

        slow, count = slow.next, 1
        while slow != fast:
            slow = slow.next
            count += 1
        return count

    def circularEntry(self, head: ListNode) -> ListNode:
        slow, fast = head.next, head.next
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
            if slow == fast:
                break

        if not fast or not fast.next:
            return None

        fast = head.next
        while slow != fast:
            slow, fast = slow.next, fast.next
        return slow
